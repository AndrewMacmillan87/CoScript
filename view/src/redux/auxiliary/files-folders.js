import axios from "axios";

const config = {
    headers: {
        "Content-Type": "application/json",
    },
};

/**
 * Gets all folders that are associated with
 * this project
 *
 * @param {String} projectId - id of the project
 */
export const getAllProjectFolders = async (projectId) => {
    try {
        const res = await axios.get(`/api/folders/${projectId}`);
        return res.data;
    } catch (err) {
        return [];
    }
};

/**
 * Creates a new instance of ProjectFolder in
 * the database and associates it with this
 * project
 *
 * @param {String} projectId - id of the project
 * @param {String} label     - name of the folder
 */
export const createNewFolder = async (projectId, label) => {
    try {
        const body = JSON.stringify({ projectId, label });
        const res = await axios.post("/api/folders/add", body, config);
        return res.data;
    } catch (err) {
        return false;
    }
};

/**
 * Creates a new instance of File in the
 * database and associates it with this project
 *
 * @param {String} folderId  - id of the parent folder
 * @param {String} label     - name of the file
 */
export const createNewFile = async (folderId, label) => {
    try {
        const body = JSON.stringify({ label });
        const res = await axios.post(
            `/api/files/add/${folderId}`,
            body,
            config
        );
        return res.data;
    } catch (err) {
        return false;
    }
};
