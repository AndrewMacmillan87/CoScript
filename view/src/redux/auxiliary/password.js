const axios = require("axios");

/**
 * Compares a newly created password against
 * all items in a user's passwordHistory property
 * in the database
 *
 * @param {Object} user     - logged-in user
 * @param {String} password - new password
 */
export const comparePW = async (user, password) => {
    if (!password) return false;
    try {
        const config = {
            headers: {
                "Content-Type": "application/json",
            },
        };
        const body = JSON.stringify({ user, password });
        const res = await axios.post("/api/changepassword/check", body, config);
        return res.data.clash;
    } catch (err) {
        return true;
    }
};

/**
 * Checks the user's current password to see if it
 * is correct before allowing the user to set a new
 * password
 *
 * @param {Object} user     - logged-in user
 * @param {String} password - user's current password
 */
export const checkCurrentPassword = async (user, password) => {
    if (!password) return false;
    try {
        const config = {
            headers: {
                "Content-Type": "application/json",
            },
        };
        const body = JSON.stringify({ user, password });
        const res = await axios.post(
            "/api/changepassword/check/current",
            body,
            config
        );
        return res.data.correct;
    } catch (err) {
        return false;
    }
};
