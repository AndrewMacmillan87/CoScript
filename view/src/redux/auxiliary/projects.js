/**
 * Shuffles an array of projects randomly, in
 * place, to be displayed in the index page
 * carousel menu
 *
 * @param {Array} array - an array of projects
 */
export const shuffle = (array) => {
    if (array.length > 2) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
    }
};

/**
 * Breaks up a larger array of projects into a 2D
 * matrix of project array, each of length <width>.
 * Any remaining projects that equate to less than
 * <width> are added at the end of the 2D matrix
 *
 * @param {Array} array  - an array of projects
 * @param {Number} width - number of projects to show
 */
export const segment = (array, width) => {
    const segmentedArray = [];
    if (!array || !width) return segmentedArray;
    let tempArray = [];
    for (let i = 0; i < array.length; i++) {
        tempArray.push(array[i]);
        if (tempArray.length === width) {
            segmentedArray.push(tempArray);
            tempArray = [];
        }
    }
    if (tempArray.length > 0) segmentedArray.push(tempArray);
    return segmentedArray;
};

/**
 * Filters out projects belonging to this user from
 * an array of all projects
 *
 * @param {Array} array   - an array of projects
 * @param {String} userId - id of logged-in user
 */
export const filterUsersProjects = (array, userId) => {
    if (array.length > 0) {
        return array.filter((el) => el.user === userId, []);
    }
};

/**
 * Filters out all active projects from an array
 * of projects
 *
 * @param {Array} array - an array of projects
 */
export const filterActive = (array) => {
    if (array.length > 0) {
        return array.filter((el) => !el.published, []);
    }
};

/**
 * Filters out all published projects from an array
 * of projects
 *
 * @param {Array} array - an array of projects
 */
export const filterPublished = (array) => {
    if (array.length > 0) {
        return array.filter((el) => el.published, []);
    }
};

/**
 * Loops over an array of projects to find any projects whose
 * subjects (author, title, genres) contain a matching sub-
 * string to the <searchCriteria>. Returns an array of projects
 * or an empty array if none were matched
 *
 * @param {Array} array           - an array of projects
 * @param {String} status         - the project's published status
 * @param {String} searchCriteria - the search term
 * @param {String} subject        - the field to search in
 */
export const filterByCriteria = (array, status, searchCriteria, subject) => {
    const cleanedSearch = searchCriteria.toLowerCase();
    const results = [];
    for (let project of array) {
        const condition =
            status === "active" ? !project.published : project.published;
        if (
            project[subject].toLowerCase().includes(cleanedSearch) &&
            condition
        ) {
            results.push(project);
        }
    }
    return results;
};
