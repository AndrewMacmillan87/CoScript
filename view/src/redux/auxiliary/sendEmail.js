const axios = require("axios");

const config = {
    headers: {
        "Content-Type": "application/json",
    },
};

/**
 * Sends an email to a user
 *
 * @param {String} email - user's email
 * @param {String} subject - the email subject
 * @param {String} source - process that initiated
 *                          the email being sent
 */
export const sendEmail = async (email, subject, source) => {
    let data;
    await axios
        .post(
            "/api/email/send",
            JSON.stringify({ email, subject, source }),
            config
        )
        .then((result) => (data = result))
        .catch((error) => console.log(error));
    return data;
};

/**
 * Send the user an email with a code they can use to
 * access a section of their account where they can
 * change their password
 *
 * @param {String} email - user's email
 */
export const sendPWEmail = async (email) => {
    let data;
    await axios
        .post("/api/email/send/changepw", JSON.stringify({ email }), config)
        .then((result) => (data = result))
        .catch((error) => console.log(error));
    return data;
};

/**
 * Verify the code a user has entered in order to
 * be granted access to the section of the application
 * that let's them change their password
 *
 * @param {Object} param0 - object containing the codes
 *                          needed to start changing a
 *                          user's password
 */
export const verifyPWCode = async ({
    originalCode,
    confirmCode,
    forgottenDetails,
}) => {
    const res = await axios.post(
        "/api/email/verify",
        JSON.stringify({ originalCode, confirmCode, forgottenDetails }),
        config
    );
    if (res.data.success) {
        return { success: true, msg: "" };
    }
    return { success: false, msg: "Code does not match" };
};

/**
 * Sends a username reminder and a temporary
 * password to a user who has forgotten their
 * login credentials. This email is only sent
 * after they have correctly entered the code
 * sent to them when they requested to change
 * their password
 *
 * @param {String} email -user's email
 */
export const sendTempCredentialsEmail = async (email) => {
    const res = await axios.post(
        "/api/email/send/reset",
        JSON.stringify({ email }),
        config
    );
    return res.data.success;
};
