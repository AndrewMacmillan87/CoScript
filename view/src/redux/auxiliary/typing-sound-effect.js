const typingSoundEffect = require("../../sound/320571__mlekoman__type-writing.wav");
const audioEffect = new Audio(typingSoundEffect);

/**
 * Starts playing the typewriter sound
 * effect
 */
export const playSoundEffect = () => {
    audioEffect.play();
};

/**
 * Stops playing the typewriter sound
 * effect
 */
export const stopSoundEffect = () => {
    audioEffect.pause();
    audioEffect.currentTime = 0;
};
