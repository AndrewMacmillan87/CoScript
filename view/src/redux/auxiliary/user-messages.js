import axios from "axios";

/**
 * Gets all of a users personal profile messages
 */
export const getUserMessages = async () => {
    try {
        const res = await axios.get("/api/messages");
        return res.data;
    } catch (err) {
        return [];
    }
};

/**
 * Returns whether or not a user is an admin
 * user on this project
 *
 * @param {String} projectId - id of the project
 */
export const isAdmin = async (projectId) => {
    try {
        const config = {
            headers: {
                "Content-Type": "application/json",
            },
        };
        const body = JSON.stringify({ projectId });
        const res = await axios.post("/api/messages/is-admin", body, config);
        return res.data.admin;
    } catch (err) {
        return false;
    }
};
