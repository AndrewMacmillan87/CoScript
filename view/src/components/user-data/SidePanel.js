import React, { Fragment, useEffect, useState } from "react";
import { Link, withRouter } from "react-router-dom";

import { getUserMessages } from "../../redux/auxiliary/user-messages";

// Redux
import { connect } from "react-redux";

import PropTypes from "prop-types";

import "./side-panel.css";

export const SidePanel = withRouter(({ auth, ...props }) => {
    const [unreadMessages, setUnreadMessages] = useState(false);

    useEffect(() => {
        const check = async () => {
            const userMessages = await getUserMessages();
            for (let message of userMessages) {
                if (!message.read) {
                    setUnreadMessages(true);
                    return;
                }
            }
            setUnreadMessages(false);
        };
        check();
    }, []);

    const user = require("../../img/user.svg");

    const profile = (
        <aside className='side-panel-container'>
            <div className='profile-avatar-container'>
                <div>
                    <img
                        className='profile-avatar'
                        src={auth.user ? auth.user.avatar : user}
                        alt='Users profile'
                    />
                </div>
                <p className='avatar-caption'>
                    <span style={{ fontWeight: "bold" }}>
                        {auth.user ? auth.user.username : ""}
                    </span>
                </p>
            </div>
            <div className='profile-options-container'>
                <ul className='profile-options'>
                    <a href='#profile-h2'>
                        <li className='profile-choice'>Profile</li>
                    </a>
                    <a href='#photo-h2'>
                        <li className='profile-choice'>Photo</li>
                    </a>
                    <a href='#projects-h2'>
                        <li className='profile-choice'>My Projects</li>
                    </a>
                    {unreadMessages ? (
                        <div className='unread-message-link'>
                            <div className='unread-message-icon'></div>
                            <Link to='/messages'>
                                <li className='profile-choice'>Messages</li>
                            </Link>
                        </div>
                    ) : (
                        <Link to='/messages'>
                            <li className='profile-choice'>Messages</li>
                        </Link>
                    )}
                    <Link to='/account'>
                        <li className='profile-choice'>Account</li>
                    </Link>
                    <Link to='/'>
                        <li className='profile-choice'>Back Home</li>
                    </Link>
                    <a
                        onClick={() => {
                            props.history["from"] = "logout";
                            props.history.push("/redirect");
                        }}
                        href=' '>
                        <li className='profile-choice'>Sign Out</li>
                    </a>
                </ul>
            </div>
        </aside>
    );

    const account = (
        <aside className='side-panel-container'>
            <div className='profile-avatar-container'>
                <div>
                    <img
                        className='profile-avatar'
                        src={auth.user ? auth.user.avatar : user}
                        alt='Users profile'
                    />
                </div>
                <p className='avatar-caption'>
                    {" "}
                    <span style={{ fontWeight: "bold" }}>
                        {auth.user ? auth.user.username : ""}
                    </span>
                </p>
            </div>
            <div className='profile-options-container'>
                <ul className='profile-options'>
                    <Link to='/profile'>
                        <li className='profile-choice'>Profile</li>
                    </Link>
                    <a href='#account-h2'>
                        <li className='profile-choice'>Account</li>
                    </a>
                    <a href='#privacy-h2'>
                        <li className='profile-choice'>Privacy</li>
                    </a>
                    <Link to='/'>
                        <li className='profile-choice'>Back Home</li>
                    </Link>
                    <a
                        onClick={() => {
                            props.history["from"] = "logout";
                            props.history.push("/redirect");
                        }}
                        href=' '>
                        <li className='profile-choice'>Sign Out</li>
                    </a>
                </ul>
            </div>
        </aside>
    );

    const messages = (
        <aside className='side-panel-container'>
            <div className='profile-avatar-container'>
                <div>
                    <img
                        className='profile-avatar'
                        src={auth.user ? auth.user.avatar : user}
                        alt='Users profile'
                    />
                </div>
                <p className='avatar-caption'>
                    {" "}
                    <span style={{ fontWeight: "bold" }}>
                        {auth.user ? auth.user.username : ""}
                    </span>
                </p>
            </div>
            <div className='profile-options-container'>
                <ul className='profile-options'>
                    <Link to='/profile'>
                        <li className='profile-choice'>Profile</li>
                    </Link>
                    <Link to='/account'>
                        <li className='profile-choice'>Account</li>
                    </Link>
                    <Link to='/'>
                        <li className='profile-choice'>Back Home</li>
                    </Link>
                    <a
                        onClick={() => {
                            props.history["from"] = "logout";
                            props.history.push("/redirect");
                        }}
                        href=' '>
                        <li className='profile-choice'>Sign Out</li>
                    </a>
                </ul>
            </div>
        </aside>
    );

    switch (props.location.pathname) {
        case "/profile":
            return profile;
        case "/account":
            return account;
        case "/messages":
            return messages;
        default:
            return <Fragment />;
    }
});

SidePanel.propTypes = {
    auth: PropTypes.object,
};

const setProps = (state) => ({
    auth: state.auth,
});

export default connect(setProps, {})(SidePanel);
