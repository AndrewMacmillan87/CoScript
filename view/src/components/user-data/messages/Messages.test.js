import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../../redux/store/reduxStore";

import { Messages } from "./Messages";
import MobileNav from "../MobileNav";
import SidePanel from "../SidePanel";

import * as auxFuncs from "../../../redux/auxiliary/user-messages";

const messageData = require("./test-data/messages.json");
const originalGetUserMessages = auxFuncs.getUserMessages;
const originalIsAdmin = auxFuncs.isAdmin;

describe("Messages component", () => {
    let props, mockSendMessage, mockSetMessageRead, mockAddCollaborator, mockLoadUserProfile;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockSendMessage = jest.fn();
        mockSetMessageRead = jest.fn();
        mockAddCollaborator = jest.fn();
        mockLoadUserProfile = jest.fn();
        props = {
            user: { userId: "123456abcdefghijklmnopqrstuvwxyz" },
            userProfile: { id: "123456" },
            sendMessage: mockSendMessage,
            setMessageRead: mockSetMessageRead,
            addCollaborator: mockAddCollaborator,
            loadUserProfile: mockLoadUserProfile,
            addFriend: jest.fn(),
            sendDeclineMsg: jest.fn(),
            blockUser: jest.fn(),
            history: {
                from: "",
                push: jest.fn(),
            },
        };
    });

    afterEach(() => {
        auxFuncs.getUserMessages = originalGetUserMessages;
        auxFuncs.isAdmin = originalIsAdmin;
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Messages {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a main element with a className of 'profile-container'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Messages {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("main").length).toEqual(1);
        expect(wrapper.find(".profile-container").length).toEqual(1);
    });

    it("should render the MobileNav component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Messages {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(MobileNav).length).toEqual(1);
    });

    it("should render the SidePanel component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Messages {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(SidePanel).length).toEqual(1);
    });

    it("should render a div with a className of 'user-info-container'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Messages {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("div").length).toEqual(3);
        expect(wrapper.find(".user-info-container").length).toEqual(1);
    });

    it("should render 1 h1 element", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Messages {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("h1").length).toEqual(1);
    });

    describe("loading", () => {
        it("should render a div element with a className of 'message-spinner-container' if 'loading' is true", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("div").length).toEqual(3);
            expect(wrapper.find(".message-spinner-container").length).toEqual(1);
        });

        it("should render an img element with a className of 'message-spinner' if 'loading' is true", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("img").length).toEqual(2);
            expect(wrapper.find(".message-spinner").length).toEqual(1);
        });

        it("should render a p element with a className of ''message-loading-msg' if 'loading' is true", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("p").length).toEqual(6);
            expect(wrapper.find(".message-loading-msg").length).toEqual(1);
        });
    });

    describe("loaded", () => {
        it("should render a section element with a className of 'profile'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find(Messages).setState({ loading: false });
            expect(wrapper.find("section").length).toEqual(1);
            expect(wrapper.find(".profile").length).toEqual(1);
        });

        it("should render a div element with a className of 'user-messages'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find(Messages).setState({ loading: false });
            expect(wrapper.find("div").length).toEqual(3);
            expect(wrapper.find(".user-messages").length).toEqual(1);
        });

        it("should render a p element with a className of 'empty-message'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find(Messages).setState({ loading: false });
            expect(wrapper.find("p").length).toEqual(6);
            expect(wrapper.find(".empty-message").length).toEqual(1);
        });

        it("should call `getUserMessages()` on mount", async () => {
            auxFuncs.getUserMessages = jest.fn();
            mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Messages {...props} />
                    </Router>
                </Provider>
            );
            await expect(auxFuncs.getUserMessages).toBeCalled();
        });

        describe("user has messages", () => {
            it("should render div elements with a className of 'user-message-container'", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({
                    loading: false,
                    messages: messageData.messages,
                });
                expect(wrapper.find(".user-message-container").length).toEqual(2);
            });

            it("should call `setMessageRead()` when the message reveal button is clicked", async () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({
                    loading: false,
                    messages: messageData.messages,
                });
                wrapper.find(".reveal-msg-btn").at(0).simulate("click");
                await expect(mockSetMessageRead).toBeCalled();
            });

            it("should call `loadUserProfile()` when the user profile link is clicked", async () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                const id = messageData.messages[0].id;
                wrapper.find(Messages).setState({
                    loading: false,
                    messages: messageData.messages,
                    messageVisibilities: { [id]: true },
                });
                wrapper.find(".sender-link").at(0).simulate("click");
                await expect(mockLoadUserProfile).toBeCalled();
            });
        });

        describe("revealReplyWindow is true", () => {
            it("should render a div element with a className of 'reply-message-box'", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealReplyWindow: true });
                expect(wrapper.find(".reply-message-box").length).toEqual(1);
            });

            it("should render a div element with a className of 'reply-message-box'", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealReplyWindow: true });
                expect(wrapper.find(".reply-message-box").length).toEqual(1);
            });

            it("should render a form element", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealReplyWindow: true });
                expect(wrapper.find("form").length).toEqual(1);
            });

            it("should render 2 button elements", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealReplyWindow: true });
                expect(wrapper.find("button").length).toEqual(2);
            });

            it("should call `sendMessage()` when the form is submitted", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealReplyWindow: true });
                wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
                expect(mockSendMessage).toBeCalled();
            });
        });

        describe("revealCollaborateWindow is true", () => {
            it("should render a div element with a className of 'collaborate-box'", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealCollaborateWindow: true });
                expect(wrapper.find(".collaborate-box").length).toEqual(1);
            });

            it("should render a p element with a className of 'collaborate-heading'", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealCollaborateWindow: true });
                expect(wrapper.find(".collaborate-heading").length).toEqual(1);
            });

            it("should render a form element", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealCollaborateWindow: true });
                expect(wrapper.find("form").length).toEqual(1);
            });

            it("should render 2 button elements", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealCollaborateWindow: true });
                expect(wrapper.find("button").length).toEqual(2);
            });

            it("should render an input element", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealCollaborateWindow: true });
                expect(wrapper.find("input").length).toEqual(1);
            });

            it("should call `addCollaborator()` when the 'Collaborate' button is clicked", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealCollaborateWindow: true });
                wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
                expect(mockAddCollaborator).toBeCalled();
            });
        });

        describe("revealFriendWindow is true", () => {
            it("should render a div element with a className of 'collaborate-box'", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealFriendWindow: true });
                expect(wrapper.find(".collaborate-box").length).toEqual(1);
            });

            it("should render a form element", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({ revealFriendWindow: true });
                expect(wrapper.find("form").length).toEqual(1);
            });

            it("should render the accept message with the 'accept' actionName", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({
                    revealFriendWindow: true,
                    actionName: "accept",
                    friendName: "Test Name",
                });

                const first = wrapper.find(".collaborate-heading").prop("children")[1].props.children[0];
                const second = wrapper.find(".collaborate-heading").prop("children")[1].props.children[1];
                expect(wrapper.find(".collaborate-heading").prop("children").toString()).toStrictEqual(
                    [
                        "Accept ",
                        <b>
                            {first}
                            {second}
                        </b>,
                        " request and add as a friend?",
                    ].toString()
                );
            });

            it("should render the decline message with the 'decline' actionName", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({
                    revealFriendWindow: true,
                    actionName: "decline",
                    friendName: "Test Name",
                });

                const first = wrapper.find(".collaborate-heading").prop("children")[1].props.children[0];
                const second = wrapper.find(".collaborate-heading").prop("children")[1].props.children[1];
                expect(wrapper.find(".collaborate-heading").prop("children").toString()).toStrictEqual(
                    [
                        "Decline ",
                        <b>
                            {first}
                            {second}
                        </b>,
                        " request?",
                    ].toString()
                );
            });

            it("should render the block message with the 'block' actionName", () => {
                const wrapper = mount(
                    <Provider store={reduxStore}>
                        <Router>
                            <Messages {...props} />
                        </Router>
                    </Provider>
                );
                wrapper.find(Messages).setState({
                    revealFriendWindow: true,
                    actionName: "block",
                    friendName: "Test Name",
                });

                const first = wrapper.find(".collaborate-heading").prop("children")[1].props.children[0];
                expect(wrapper.find(".collaborate-heading").prop("children").toString()).toStrictEqual(
                    ["Block ", <b>{first}</b>, "? You will no longer be able to send or receive communications to or from this user."].toString()
                );
            });
        });
    });
});
