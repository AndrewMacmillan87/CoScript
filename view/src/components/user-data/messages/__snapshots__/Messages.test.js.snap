// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages component renders correctly 1`] = `
<main
  className="profile-container"
>
  <nav
    className="profile-mobile-nav"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <i
        className="home-icon fas fa-home fa-3x"
      />
    </a>
    <button
      className="profile-dropdown-icon-link"
      onClick={[Function]}
    >
      <img
        alt="Burger menu icon"
        className="profile-dropdown-icon"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "menu.svg",
          }
        }
      />
    </button>
    <div
      className="dropdown-menu dropdown-content"
      id="dropdown-nav"
    >
      <a
        href="/profile"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          My Profile
        </p>
      </a>
      <a
        href="/messages"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Messages
        </p>
      </a>
      <a
        href="/projects"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Active Projects
        </p>
      </a>
      <a
        href="/publications"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Publications
        </p>
      </a>
      <a
        href=" "
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Logout
        </p>
      </a>
    </div>
  </nav>
  <div
    className="user-info-container"
  >
    <h1
      className="profile-h1"
    >
      My Messages
    </h1>
    <div
      className="message-spinner-container"
    >
      <img
        alt="loading"
        className="message-spinner"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "loading-black.svg",
          }
        }
      />
      <p
        className="message-loading-msg"
      >
        Loading messages...
      </p>
    </div>
  </div>
</main>
`;
